cli/unstable_prompt_select.ts

```diff
@@ -0,0 +1,76 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+/** Options for {@linkcode promptSelect}. */
+export interface PromptSelectOptions {
+  /** Clear the lines after the user's input. */
+  clear?: boolean;
+}
+
+const ETX = "\x03";
+const ARROW_UP = "\u001B[A";
+const ARROW_DOWN = "\u001B[B";
+const CR = "\r";
+const INDICATOR = "❯";
+const PADDING = " ".repeat(INDICATOR.length);
+
+const CLR = "\r\u001b[K"; // Clear the current line
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+/**
+ * Shows the given message and waits for the user's input. Returns the user's selected value as string.
+ *
+ * @param message The prompt message to show to the user.
+ * @param values The values for the prompt.
+ * @param options The options for the prompt.
+ * @returns The string that was entered or `null` if stdin is not a TTY.
+ *
+ * @example Usage
+ * ```ts ignore
+ * import { promptSelect } from "@std/cli/prompt-select";
+ *
+ * const browser = promptSelect("Please select a browser:", ["safari", "chrome", "firefox"], { clear: true });
+ * ```
+ */
+export function promptSelect(
+  message: string,
+  values: string[],
+  { clear }: PromptSelectOptions = {},
+): string | null {
+  const length = values.length;
+  let selectedIndex = 0;
+
+  Deno.stdout.writeSync(encoder.encode(`${message}\r\n`));
+  Deno.stdin.setRaw(true);
+
+  const buffer = new Uint8Array(4);
+  loop:
+  while (true) {
+    for (const [index, value] of values.entries()) {
+      const start = index === selectedIndex ? INDICATOR : PADDING;
+      Deno.stdout.writeSync(encoder.encode(`${start} ${value}\r\n`));
+    }
+    const n = Deno.stdin.readSync(buffer);
+    if (n === null || n === 0) break;
+    const input = decoder.decode(buffer.slice(0, n));
+    switch (input) {
+      case ETX:
+        return Deno.exit(0);
+      case ARROW_UP:
+        selectedIndex = (selectedIndex - 1 + length) % length;
+        break;
+      case ARROW_DOWN:
+        selectedIndex = (selectedIndex + 1) % length;
+        break;
+      case CR:
+        break loop;
+    }
+    Deno.stdout.writeSync(encoder.encode(`\x1b[1A${CLR}`.repeat(length)));
+  }
+  if (clear) {
+    Deno.stdout.writeSync(encoder.encode(`\x1b[1A${CLR}`.repeat(length + 1))); // clear values and message
+  }
```

Comment:
I personally don't remember any tool which doesn't clear the selections in this context. How about always clearing values and message?

Reply:
the built in `prompt()`doesn't clear.
This option is taken over from `promptSecret()`. I think if it is there, we should provide it here also.

Reply:
I mean I've never seen a tool which doesn't clear the choices part of the prompt. For example, `npm init vite` prompts several select-type prompts, but choices areas are always cleared after the selection. Leaving choices as is doesn't look like sensible default behavior to me.

Reply:
If we default to leave something on the screen after the selection, then I think the output should become something like:

```
Please select a browser: safari
```

instead of leaving the entire selection section as is, like:

```
Please select a browser:
❯ safari
  chrome
  firefox
```

(for the call `promptSelect("Please select a browser:", ["safari", "chrome", "firefox"])`

Reply:
I see. In that case I don't know what the best default is. I like the idea of not clearing the lines by default so the history remains and aligns with `prompt()` and `promptSecret()` but it doesn't make much sense to print a different line imo...

Reply:
hmm, ok. Let's keep this as is

Reply:
👍 since it is unstable, we can adjust according to user feedback.